solr/test-framework/src/java/org/apache/solr/util/EmbeddedSolrServerTestRule.java

```diff
@@ -0,0 +1,273 @@
+      return this;
+    }
+
+    public Path getSolrHome() {
+      return solrHome;
+    }
+
+    public String getSchemaFile() {
+      return schemaFile;
+    }
+
+    public String getConfigFile() {
+      return configFile;
+    }
+
+    public String getCollectionName() {
+      return collectionName;
+    }
+
+    public RequestWriterSupplier getRequestWriterSupplier() {
+      return requestWriterSupplier;
+    }
+
+    public void init() {
+
+      EmbeddedSolrServerTestRule.this.init(this);
+    }
+  }
+
+  private void init(Builder b) {
+    Path solrHome = b.getSolrHome();
+    String schemaFile = b.getSchemaFile();
+    String configFile = b.getConfigFile();
+    String collectionName = b.getCollectionName();
+
+    if (b.getRequestWriterSupplier() != null) {
+      SolrConfig solrConfig;
+
+      try {
+        solrConfig = new SolrConfig(solrHome.resolve(collectionName), configFile);
+      } catch (Exception e) {
+        throw new RuntimeException(e);
+      }
+
+      NodeConfig nodeConfig = buildTestNodeConfig(solrHome);
+
+      TestCoresLocator testCoreLocator =
+          new TestCoresLocator(
+              collectionName,
+              LuceneTestCase.createTempDir("data-dir").toFile().getAbsolutePath(),
+              solrConfig.getResourceName(),
+              IndexSchemaFactory.buildIndexSchema("schema.xml", solrConfig).getResourceName());
+
+      CoreContainer container = new CoreContainer(nodeConfig, testCoreLocator);
+      container.load();
+      client = new EmbeddedSolrServer(container, collectionName, b.getRequestWriterSupplier());
+    } else if (schemaFile != null) {
+      SolrConfig solrConfig;
+
+      try {
+        solrConfig = new SolrConfig(solrHome.resolve(collectionName), configFile);
+      } catch (Exception e) {
+        throw new RuntimeException(e);
+      }
+
+      NodeConfig nodeConfig = buildTestNodeConfig(solrHome);
+
+      TestCoresLocator testCoreLocator =
+          new TestCoresLocator(
+              collectionName,
+              LuceneTestCase.createTempDir("data-dir").toFile().getAbsolutePath(),
+              solrConfig.getResourceName(),
+              IndexSchemaFactory.buildIndexSchema(schemaFile, solrConfig).getResourceName());
+
+      CoreContainer container = new CoreContainer(nodeConfig, testCoreLocator);
+      container.load();
+      client = new EmbeddedSolrServer(container, collectionName);
+
+    } else {
+      client = new EmbeddedSolrServer(solrHome, collectionName);
+    }
+  }
+
+  public Builder build() {
+    return new Builder();
+  }
+
+  @Override
+  protected void before() throws Throwable {
+    super.before();
+  }
+
+  @Override
+  protected void after() {
+
+    try {
+
+      client.close();
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+    client.getCoreContainer().shutdown();
+  }
+
+  @Override
+  public EmbeddedSolrServer getSolrClient() {
+    return client;
```

Comment:
I would assert that the client isn't null.  I doubt any test checks for null-ness; they assume it's set.